lib/ufe/UsdAttribute.cpp

```diff
@@ -47,6 +48,36 @@ static constexpr char kErrorMsgEnumNoValue[] = "Enum string attribute has no val
 namespace
 {
 
+template<typename T>
+bool setUsdAttr(const PXR_NS::UsdAttribute& attr, const T& value)
+{
+    // As of 24-Nov-2019, calling Set() on a UsdAttribute causes two "info only"
+    // change notifications to be sent (see StagesSubject::stageChanged).  With
+    // the current USD implementation (USD 19.11), UsdAttribute::Set() ends up
+    // in UsdStage::_SetValueImpl().  This function calls in sequence:
+    // - UsdStage::_CreateAttributeSpecForEditing(), which has an SdfChangeBlock
+    //   whose expiry causes a notification to be sent.
+    // - SdfLayer::SetField(), which also has an SdfChangeBlock whose
+    //   expiry causes a notification to be sent.
+    // These two calls appear to be made on all calls to UsdAttribute::Set(),
+    // not just on the first call.
+    // 
+    // Trying to wrap the call to UsdAttribute::Set() inside an additional
+    // SdfChangeBlock fails: no notifications are sent at all.  This is most
+    // likely because of the warning given in the SdfChangeBlock documentation:
+    // 
+    // https://graphics.pixar.com/usd/docs/api/class_sdf_change_block.html
+    //
+    // which stages that "it is not safe to use [...] [a] downstream API [such
+    // as Usd] while a changeblock is open [...]".
+    //
+    // Therefore, we have implemented an attribute change block notification of
+    // our own in the StagesSubject, which we invoke here, so that only a
+    // single UFE attribute changed notification is generated. 
```

Comment:
Would love to get Pixar feedback on this.  It feels like a single call to UsdAttribute Set() should cause a single USD notification, but we get two, as per the comment, and thus the guard workaround.

Reply:
@ppt-adsk: Something seems awry here. Your breakdown of the call sequence looks correct, and for the initial creation of an attribute, there should indeed be two notifications: one for the creation of the attribute spec, and then a second for the "info only" change in its value. Subsequent calls to `Set()` on the same attribute should only generate a single notification though. They should end up bailing out of `_CreatePropertySpecForEditing()` early, before an SdfChangeBlock is created or any new property spec is created:
https://github.com/PixarAnimationStudios/USD/blob/4b1162957b2ad219e1635c81de8343be490e41fc/pxr/usd/lib/usd/stage.cpp#L1337

This should be getting tested in core USD here:
https://github.com/PixarAnimationStudios/USD/blob/4b1162957b2ad219e1635c81de8343be490e41fc/pxr/usd/lib/usd/testenv/testUsdNotices.py#L82

Though that tests authoring metadata, I amended the test script locally and verified that the behavior is the same for regular attributes:
```
            attr = s.GetPrimAtPath("/Foo").CreateAttribute(
                "attr", Sdf.ValueTypeNames.Int)
            self.assertEqual(self._changeCount, 4)
            self.assertEqual(self._contentsCount, 2)
            self.assertEqual(self._objectsCount, 2)
            self.assertEqual(self._editTargetsCount, 0)

            self._ResetCounters()
            attr.Set(42)
            self.assertEqual(self._changeCount, 2)
            self.assertEqual(self._contentsCount, 1)
            self.assertEqual(self._objectsCount, 1)
            self.assertEqual(self._editTargetsCount, 0)
```

Reply:
Re-reading your post, and focusing on one of the aspects: "for the initial creation of an attribute, there should indeed be two notifications: one for the creation of the attribute spec, and then a second for the "info only" change in its value."  Does the initial creation of the attribute mean creation of the attribute spec in the edit target layer?  If so, can it be argued that this specific to the USD data model, and should be abstracted away by UFE, which should provide only a single notification of the attribute value being set?  A client interested in USD details should subscribe to USD notifications, and not go through UFE notifications.

I realize this is not answering the question of why subsequent calls to USD set attribute aren't generating a single notification, but I'm wondering whether we need the guard around USD set attribute regardless, because the initial set attribute is expected to generate two USD notifications, which we want to collapse down to a single UFE notification.

Reply:
Indeed, the goal for UFE is to abstract the data model. We don't want to mimic any specific behaviors one data model would have.

Reply:
I agree with you.  Hopefully we'll be able to gain insight into this through other work and come back to it with a clearer picture.